hackathon_site/dashboard/frontend/src/testing/utils.tsx

```diff
@@ -0,0 +1,79 @@
+import React from "react";
+import { BrowserRouter } from "react-router-dom";
+import { Provider } from "react-redux";
+import {
+    render as rtlRender,
+    RenderOptions as RtlRenderOptions,
+} from "@testing-library/react";
+
+import { makeStore, RootStore, RootState } from "slices/store";
+import { DeepPartial } from "redux";
+import { SnackbarProvider } from "notistack";
+import { AxiosResponse } from "axios";
+import { APIListResponse } from "api/types";
+
+export const withRouter = (component: React.ComponentElement<any, any>) => (
+    <BrowserRouter>{component}</BrowserRouter>
+);
+
+export const withStore = (
+    component: React.ComponentElement<any, any>,
+    store: RootStore = makeStore()
+) => <Provider store={store}>{component}</Provider>;
+
+export const withStoreAndRouter = (
+    component: React.ComponentElement<any, any>,
+    store: RootStore = makeStore()
+) => withStore(withRouter(component), store);
+
+interface RenderOptions {
+    preloadedState?: DeepPartial<RootState>;
+    store?: RootStore;
+    options?: Omit<RtlRenderOptions, "queries">;
+}
+
+const customRender = (
+    ui: React.ReactElement,
+    {
+        preloadedState = {},
+        store = makeStore(preloadedState),
+        ...renderOptions
+    }: RenderOptions = {}
+) => {
+    const wrapper = ({ children }: any) => (
+        <Provider store={store}>
+            <SnackbarProvider>
+                <BrowserRouter>{children}</BrowserRouter>
+            </SnackbarProvider>
+        </Provider>
+    );
+    return rtlRender(ui, { wrapper, ...renderOptions });
+};
+
+// Re-export everything from rtl
+export * from "@testing-library/react";
+
+// Override render method
+export { customRender as render };
+
+export const promiseResolveWithDelay = <T extends unknown>(
+    value: T,
+    ms: number = 500
+): Promise<T> =>
+    new Promise((resolve) => {
+        setTimeout(() => resolve(value), ms);
+    });
+
+export const makeMockApiListResponse = <T extends unknown>(
+    data: T[],
+    next?: string,
+    previous?: string
+): AxiosResponse<APIListResponse<T>> =>
```

Comment:
wondering why you utilised an AxiosResponse instead of regular Promises?

Reply:
Tldr; I wrap what this helper returns in a promise eventually. `AxiosResponse` itself is not a promise, it's just another type definition.

The return type of `get()` in `api/api.ts` is `Promise<AxiosResponse<T>>` (because that's what axios returns).

This helper function is just used to construct a list response from a set of items, so for example 

```typescript
const hardware: Hardware[] = [/* Define some hardware objects here */]
const mockListResponse = makeMockApiListResponse(hardware)
```

will give `mockListResponse` a type of `AxiosResponse<APIListResponse<Hardware>>` (note that we don't actually need to type annotate it, typescript is really good at inferring types).

Given that, the question is why return `AxiosResponse<...>` instead of `Promise<AxiosResponse<...>>`?

Because where I actually use this function, I eventually do wrap the return value in a promise.

For example, in the inventory grid tests:

```
jest.mock("api/api");
const mockedGet = get as jest.MockedFunction<typeof get>;
...

// Inside some test
const apiResponse = makeMockApiListResponse(mockHardware);
mockedGet.mockResolvedValue(apiResponse);
```

`mockResolvedValue` just means "when something calls the get function, have it return a resolved promise with this value` - so that's where the promise wrapper is coming from. 

I could have also done (and used to before I discovered `mockResolvedValue`):

```typescript
mockedGet.mockReturnValue(Promise.resolve(apiResponse));
```

which is the same thing. Note if I just tried to return `apiResponse`, typescript will actually complain since it knows that `get` should be returning `Promise<AxiosResponse<...>>`, not just `AxiosResponse<...>`.

If I really wanted to manually resolve the promise, I could do that to:

```typescript
mockedGet.mockReturnValue(new Promise((resolve) => {
    resolve(apiResponse)
}));
```

If you take a look at `promiseResolveWithDelay` in utils.tsx and where I use it, I pretty much do exactly that - resolve the promise after a delay set in a `setTimeout`.